atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -263,13 +266,18 @@ public InetSocketAddress getRandomHostForKey(byte[] key) {
         if (liveOwnerHosts.isEmpty()) {
             log.warn("Perf / cluster stability issue. Token aware query routing has failed because there are no known " +
                     "live hosts that claim ownership of the given range. Falling back to choosing a random live node. " +
-                    "For debugging, our current ring view is: {} and our current host blacklist is {}", tokenMap, blacklistedHosts);
+                    "Current state logged at DEBUG");
+            log.debug("Current ring view is: {} and our current host blacklist is {}", tokenMap, blacklistedHosts);
             return getRandomGoodHost().getHost();
         } else {
-            return currentPools.get(ImmutableList.copyOf(liveOwnerHosts).get(ThreadLocalRandom.current().nextInt(liveOwnerHosts.size()))).getHost();
+            return getRandomHostByActiveConnections(Maps.filterKeys(currentPools, liveOwnerHosts::contains));
         }
     }
 
+    private static InetSocketAddress getRandomHostByActiveConnections(Map<InetSocketAddress, CassandraClientPoolingContainer> pools) {
+        return WeightedHosts.create(pools).getRandomHost();
```

Comment:
We could profile it, but I don't think it's anywhere near as expensive as the billion other lines of code it runs. This is basically:
1) create a list of ~20 things (and hopefully most of the time it's more like 3 things if we have a specific token)
2) iterate over them, grabbing the utilization field (only requires implicit read locks)
3) iterate over them to create a map of the same size
4) pick randomly

So the only thing that would even potentially take time is getting the pool utilization. It looks like ConcurrentHashMap doesn't even grab a read lock, but LinkedBlockingDeque does. I doubt it's an issue anyways when it's done # of hosts times per network call.